url270fc4f5ee_ykcilborw_Joust_tgz-pJ8-fri_patterns_interpreter_parsergenerator_examples_HowToOverrideLexerImplJ8/src/fri/patterns/interpreter/parsergenerator/Token.java

```diff
@@ -0,0 +1,149 @@
+	public static final String TOKEN = "token";
+	
+	/** Reserved identifier that marks tokens not to deliver to parser, e.g. <i>ignored ::= spaces ;</i>. */
+	public static final String IGNORED = "ignored";
+	
+	/** The character used to mark artificial nonterminal (e.g. made from "prolog?"). Every artificial rule starts with it. */
+	public static final String ARTIFICIAL_NONTERMINAL_START_CHARACTER = "_";
+	
+	/** The syntax symbol this Token represents. */
+	public final String symbol;
+	/** The text that was scanned for this Token. */
+	public final Object text;
+	/** The start and end Address of this Token. */
+	public final Range range;
+
+
+	/** Address stores input line number (1-n), column (0-n) and character/byte offset (0-n). */
+	public static class Address implements Comparable
+	{
+		/** The line number (1-n). */
+		public final int line;
+		/** The column number (0-n). */
+		public final int column;
+		/** The character/byte offset (0-n). */
+		public final int offset;
+		
+		public Address()	{
+			this(1, 0, 0);
+		}
+		public Address(int line, int column, int offset)	{
+			this.line = line;
+			this.column = column;
+			this.offset = offset;
+		}
+		public String toString()	{
+			return line+"/"+column;
+		}
+		public boolean equals(Object o)	{
+			return offset == ((Address) o).offset;
+		}
+		public int hashCode()	{
+			return offset;
+		}
+		public int compareTo(Object o) {
+			return offset - ((Address) o).offset;
+		}
+	}
+
+	/** Range stores start and end Address of a token. */
+	public static class Range implements Comparable
+	{
+		/** The start Address of this Token, pointing to the first character/byte. */
+		public final Address start;
+		/** The end Address of this Token, one after the last character/byte. */
+		public final Address end;
+		
+		public Range(Address start, Address end)	{
+			this.start = start != null ? start : new Address();
+			this.end = end != null ? end : new Address();
+		}
+		public boolean equals(Object o)	{
+			return start.equals(((Range) o).start) && end.equals(((Range) o).end);
+		}
+		public int hashCode()	{
+			return start.hashCode() + end.hashCode();
+		}
+		public String toString()	{
+			return start+"-"+end;
+		}
+		public int compareTo(Object o) {
+			return start.compareTo(((Range) o).start) + end.compareTo(((Range) o).end);
+		}
+	}
+
+
+	public Token(String symbol, Object text, Range range)	{
+		this.symbol = symbol;
+		this.text = text;
+		this.range = range;
+	}
+
+	/**
+		Epsilon means end of input, EOF, no more bytes available.
+		@return true if passed token is the EPSILON-symbol.
+	*/
+	public static boolean isEpsilon(Token token)	{
+		return isEpsilon(token.symbol);
+	}
+
+	/**
+		Epsilon means end of input, no more bytes available.
+		@return true if Token symbol not null and is the EPSILON-symbol.
+	*/
+	public static boolean isEpsilon(String symbol)	{
+		return symbol != null && symbol == EPSILON;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: An equality operator (== or !=) is used to compare strings, which matches strings based on address.

**Fix**: Do content comparison using the equals() method to compare the values of the strings.